Use email address validator from Zend

barbushin opened the pull request:
Closes #7298

barbushin committed:
Use email address validator from Zend

mattab commented:
Feedback: 
* add a unit test for this method to show that it really closes #7298 and that it validate email addresses properly

barbushin committed:
Fix Zend\Validate using deprecated iconv_set_encoding()

barbushin committed:
Add test for Piwik::isValidEmailString()

barbushin commented:
@mattab Done

diosmosis commented on core/Piwik.php:
```diff
@@ -563,7 +563,14 @@ public static function redirectToModule($newModule, $newAction = '', $parameters
+                'mx' => false,
+                'deep' => false
+            ));
+        }
```
Instead of using a static variable here, DI should be used. You can get an object from DI via,

```
StaticContainer::get('Zend_Validate_EmailAddress'); // full class name goes here
```

If there's no definition in a PHP config file (ie, config/global.php or one of the plugin ones), the DI container will create a new instance itself, autowiring dependencies.

Since you're creating a instance w/ a specific constructor arg, you'll have to add a definition for the instance in `config/global.php`. Should be self-explanatory how to do it, let me know if it isn't.

Regarding `StaticContainer` use: since Piwik's internals are so deeply coupled, we have to use something like this (which stores the container as a global static variable), however we're getting close to removing it. In the future, if possible, try not to use it (by specifying dependencies in the constructor). In this case, since it's a static method, there's no choice but to use it (especially since it's `@api`). If this isn't clear, let me know, I'll try to clarify.

barbushin replied:
@diosmosis StaticContainer is martked as `@deprecated`.

diosmosis replied:
It was introduced as `@deprecated` because Piwik's dependencies are too hard-coded. It's taken us ~5 months or so to try to remove it, and we are still one month away from doing it. You can use it for now, though you should try to avoid where possible.

barbushin replied:
@diosmosis I'm worried, that if I put `Zend_Validate_EmailAddress` in DI, then another core/plugins developers will call email validation using `Zend_Validate_EmailAddress` from DI, instead of using `Piwik::isValidEmailString()`, and in this way Piwik will depends on Zend/Validate library. I gues this case is not about DI pattern, it's about https://en.wikipedia.org/wiki/Flyweight_pattern

But if this is a common practice for Piwik to use DI container as Flyweighter, then I can do some refactoring. Benaka, are you 100% sure about it?

diosmosis replied:
Most of what's in DI is currently not API, and when things are made API they are explicitly mentioned in docs (or should be).

As for the flyweight pattern, that normally involves more than one object. This is caching a single instance for the current request. If you don't want to use DI, you could use the transient cache, via, `Piwik\Cache::getTransientCache()`.

Whatever you do, please make sure this value can be reset during tests, and if possible, overridden (though I think this is only do-able through DI). Static local variables have made me waste days debugging tests, and I would gladly anger the design pattern gods or whatever entity is responsible for programming zealotry if it meant I didn't have to waste my time on pointless work.

diosmosis commented on tests/PHPUnit/Integration/EmailValidatorTest.php:
```diff
@@ -0,0 +1,140 @@
+        return Piwik::isValidEmailString($email);
+    }
+
+    public function testValidStandard()
```
The current convention for test names is `test_{{methodName}}_{{descriptionOfTest}}`, for example:

`test_isValid_shouldValidateANormalEmail`

Can you change the test method names to this type of name?

More examples: https://github.com/piwik/piwik/blob/master/tests/PHPUnit/Integration/TrackerTest.php

diosmosis commented on tests/PHPUnit/Integration/EmailValidatorTest.php:
```diff
@@ -0,0 +1,140 @@
+/**
+ * @group Core
+ */
+class EmailValidatorTest extends IntegrationTestCase
```
Is there a reason this is an integration test? Note: currently the IntegrationTestCase base class will create a database before tests start and reset it before each test, which slows things down.

barbushin committed:
EmailValidatorTest inheritance change and methods renaming

barbushin commented:
@diosmosis I've renamed test methods, and changed test parent class to `PHPUnit_Framework_TestCase`. Thanks for that review!

barbushin committed:
Move Zend_Validate_EmailAddress to DI

barbushin commented:
Moved `Zend_Validate_EmailAddress` to `DI`. Now it's ready to merge?

barbushin committed:
Fix UsersManagerTest::testAddUserWrongEmail()
Mow email validation with different email formats is tested in EmailValidatorTest. UsersManagerTest::testAddUserWrongEmail() test is just about exception throwing.

barbushin commented:
It was a wrong old test method `UsersManagerTest::testAddUserWrongEmail()` that  asserted that `ema'il@email.com` is invalid, but regarding RFC and new email validator(and http://isemail.info) `ema'il@email.com` is valid.

`array('mx' => false, 'deep' => false)` are default settings for `Zend_Validate_EmailAddress` so I think we can just forget about it.

mnapoli commented on config/global.php:
```diff
@@ -64,5 +64,7 @@
-    'Piwik\EventDispatcher' => DI\object()->constructorParameter('observers', DI\get('observers.global'))
+    'Piwik\EventDispatcher' => DI\object()->constructorParameter('observers', DI\get('observers.global')),
+
+    'Zend_Validate_EmailAddress' => DI\object('Zend_Validate_EmailAddress'),
```
FYI this line is not necessary: by default with [autowiring](http://php-di.org/doc/autowiring.html) PHP-DI can create the instance without configuration needed (because here there is no constructor parameter to configure). It's no big deal but it's just to let you know

mnapoli replied:
Ah I just saw d4a024b02b0e5f18b7c88b965d836e8c609bdec9 ;)
